pandas/core/indexes/base.py

```diff
@@ -2566,6 +2567,7 @@ def _union(self, other, sort):
                 # worth making this faster? a very unusual case
                 value_set = set(lvals)
                 result.extend([x for x in rvals if x not in value_set])
+                result = Index(result)._values  # do type inference here
```

Comment:
why does _wrap_setop_result not do this?

Reply:
ATM it does in a roundabout manner.  Index._wrap_setop_result calls self._constructor, which is `type(self)` whereas the subclasses use _shallow_copy.

Reply:
ok